// @visual-regression-tracker/sdk-js/dist/__data__/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testRunAutoApprovedResponse = exports.testRunNewResponse = exports.testRunOkResponse = exports.testRunUnresolvedResponse = void 0;
const types_1 = require("../types");
exports.testRunUnresolvedResponse = {
    url: "url",
    status: types_1.TestStatus.unresolved,
    pixelMisMatchCount: 12,
    diffPercent: 0.12,
    diffTollerancePercent: 0,
    id: "some id",
    imageName: "imageName",
    merge: false,
};
exports.testRunOkResponse = {
    url: "url",
    status: types_1.TestStatus.ok,
    diffPercent: 0,
    diffTollerancePercent: 0,
    id: "some id",
    imageName: "imageName",
    merge: false,
};
exports.testRunNewResponse = {
    url: "url",
    status: types_1.TestStatus.new,
    pixelMisMatchCount: 0,
    diffPercent: 0,
    diffTollerancePercent: 0,
    id: "some id",
    imageName: "imageName",
    merge: false,
};
exports.testRunAutoApprovedResponse = {
    url: "url",
    status: types_1.TestStatus.autoApproved,
    pixelMisMatchCount: 0,
    diffPercent: 0,
    diffTollerancePercent: 0,
    id: "some id",
    imageName: "imageName",
    merge: false,
};
